### DISCUSSION TITLE
Noob Question: Differences between httpx and httpx-ws

### DISCUSSION THREAD

**Discussion by FrankC01** [frankie567/httpx-ws#17] [Q&A] [answered]
So I've been using `httpx` for both synchronous and asynchronous 'post' processing against Sui blockchain. [pysui](https://github.com/FrankC01/pysui)

They have a subscription capability for events. Looking through the httpx docs there is a notion of stream but I don't get the impression that it fully support websocket.

So my questions are:
1. Are httpx and httpx-ws complimentary? 
2. In context of httpx-ws:
    1. What is the right approach to leverage httpx when using httpx-ws?
    2. I want to support multiple websocket connections, as the user may apply different filters that will continuously send the filtered data to that connection. Is that possible?
   
   Sorry for the newbie questions...

**Comment by frankie567** [comment, marked as answer]
1. The purpose of HTTPX is to make HTTP requests (GET, POST, PUT...) and receive the response, whether it's HTML, JSON or whatever.
2. HTTPX-WS is a tool built on top of HTTPX to help you talk with WebSockets. It relies on HTTPX to perform the initial connection and then gives you access to a set of useful methods to send and receive data from a WebSocket.

So, they are complimentary in the sense that they don't serve the same purpose. In your client implementation, you'll probably have two approaches:

* The HTTP approach: I call a function, I get a result, e.g:

```py
data = sui_client.get_transaction("TRANSACTION_ID")
print(data)
```

* The WebSocket approach: I open a connection, I receive events in real-time until I close the connection, e.g:

```py
with sui_client.listen_events(filter="sender:SENDER_ID") as stream:
    while True:
        data = stream.receive()  # Will block until a message is received
        print(data)
```

In the context of a client, I don't really understand what you want to mean by "support multiple websocket connections". IMO, you should provide the required function (like above) to open an event stream with a specific filter.

If users of your client wants to listen for several streams at the same time, well, that's up to them to implement this (maybe they'll do multi-threading, or dispatch to a broker, there a lot of solutions for this kind of thing).
